api/src/main/java/org/apache/iceberg/actions/DeleteOrphanFiles.java

```diff
@@ -80,6 +80,19 @@
    */
   DeleteOrphanFiles executeDeleteWith(ExecutorService executorService);
 
+  /**
+   * Passes a table which contains the list of actual files in the table. This skips the directory listing - any
+   * files in the actualFilesTable provided which are not found in table metadata will be deleted. Not compatible
+   * with `location` or `older_than` arguments - this assumes that the provided table of actual files has been
+   * filtered down to the table’s location and only includes files older than a reasonable retention interval.
+   *
+   * @param tableName the table containing the actual files dataset.  Should have a single `file_path` string column
+   * @return this for method chaining
+   */
+  default DeleteOrphanFiles actualFilesTable(String tableName) {
```

Comment:
I think it is a good idea to be able to use a temporary table for the stored procedure, but for the action API I would expect to be able to pass in a `Dataset` of some kind. `Dataset<String>` would work, and we could also expose a couple of Java classes to make `Dataset<FileLocationAndModifiedTime>` and `Dataset<FileLocation>` work.

I think that we can also make what's happening here a bit more clear in the API by naming the method something like `compareToFileList(Dataset<String>)`:

```java
Dataset<String> files = ...
SparkActions.get()
    .deleteOrphanFiles(catalog.loadTable(identifier))
    .compareToFileList(files)
    .execute();
```

Reply:
I guess the problem is that the Action API is engine agnostic so we cannot refer to `Dataset`. We considered offering engine-specific actions at some point but that idea was discarded.

Reply:
I think we should move this into the Spark-specific API then.

Reply:
@rdblue, you mean exposing this method in `BaseDeleteOrphanFilesSparkAction` similarly to `expire` in `BaseExpireSnapshotsSparkAction` or introduce a Spark-specific interface?

It is a bummer as this functionality that can be useful for all engines.

Reply:
We may be able to do this later, but I think the most natural way to do this for the Actions API is to use a Dataset. We can take another look at this later, too. Maybe we should parameterize this method so that you can specify what collection you want. Flink could use DataStream and Spark could use Dataset.